python/lsst/ctrl/bps/clustered_quantum_graph.py

```diff
@@ -23,65 +23,473 @@
+            a cluster with given name.
         """
-        self.add_node(name, qgraph_node_ids=node_ids, label=label, tags=tags)
+        if name not in self._cluster_graph:
+            raise KeyError(f"{self.name} does not have a cluster named {name}")
 
-    def add_node(self, node_for_adding, **attr):
-        """Override add_node function to ensure that nodes are limited
-        to QuantumGraphs.
+        _LOG.debug("get_cluster nodes = %s", list(self._cluster_graph.nodes))
+        attr = self._cluster_graph.nodes[name]
+        return attr['cluster']
+
+    def get_quantum_node(self, id):
+        """Retrieve a QuantumNode from the ClusteredQuantumGraph by ID.
 
         Parameters
         ----------
-        node_for_adding : `str` or `list` [`lsst.pipe.base.NodeId`]
-            Name of cluster or cluster data (list of NodeIds).
-        attr : keyword arguments, optional
-            Attributes to be saved with node in graph.
-        """
-        if isinstance(node_for_adding, list):
-            name = f"{len(self) + 1:06d}"  # Create cluster name by counter.
-            attr["qgraph_node_ids"] = node_for_adding
-        elif "qgraph_node_ids" in attr:
-            if not isinstance(attr["qgraph_node_ids"], list):
-                raise RuntimeError(f"Invalid type {type(attr['qgraph_node_ids'])}"
-                                   " for qgraph_node_ids attribute; should be list.")
-            name = node_for_adding
+        id : `lsst.pipe.base.NodeId` or int
+            ID of the QuantumNode to retrieve.
+
+        Returns
+        -------
+        quantum_node : `lsst.pipe.base.QuantumNode`
+            QuantumNode matching given ID.
+
+        Raises
+        ------
+        KeyError
+            Raised if the ClusteredQuantumGraph does not contain
+            a QuantumNode with given ID.
+        """
+        node_id = id
+        if isinstance(id, int):
+            node_id = NodeId(id, self._quantum_graph.graphID)
+        _LOG.debug("get_quantum_node: node_id = %s", node_id)
+        return self._quantum_graph.getQuantumNodeByNodeId(node_id)
+
+    def __iter__(self):
+        """Iterate over names of clusters.
+
+        Returns
+        -------
+        names : `Iterator` [`str`]
+            Iterator over names of clusters.
+        """
+        return self._cluster_graph.nodes()
+
+    def clusters(self):
+        """Iterate over clusters.
+
+        Returns
+        -------
+        clusters : `Iterator` [`lsst.ctrl.bps.QuantaCluster`]
+            Iterator over clusters.
+        """
+        return map(self.get_cluster, self._cluster_graph.nodes())
+
+    def successors(self, name):
+        """Return clusters that are successors of the cluster
+        with the given name.
+
+        Parameters
+        ----------
+        name : `str`
+            Name of cluster for which need the successors.
+
+        Returns
+        -------
+        clusters : `Iterator`
```

Comment:
It would be nice to put in the description (here and in the following `predecessor` function) what `clusters` iterates over (`lsst.ctrl.bps.QuantaCluster`s I assume).